man/vcfroc.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "VCFROC" "1" "" "vcfroc (vcflib)" "vcfroc (VCF statistics)"
.hy
.SH NAME
.PP
\f[B]vcfroc\f[R]
.SH SYNOPSIS
.PP
\f[B]vcfroc\f[R] options []
.SH DESCRIPTION
.PP
Generates a pseudo-ROC curve using sensitivity and specificity estimated
against a putative truth set.
Thresholding is provided by successive QUAL cutoffs.
.SH OPTIONS
.IP
.nf
\f[C]

options:
    -t, --truth-vcf FILE      use this VCF as ground truth for ROC generation
    -w, --window-size N       compare records up to this many bp away (default 30)
    -c, --complex             directly compare complex alleles, don\[aq]t parse into primitives
    -r, --reference FILE      FASTA reference file

\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.

src/vcfroc.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

#include "Variant.h"
#include "legacy.h"
#include "BedReader.h"
#include "IntervalTree.h"
#include <getopt.h>
#include "Fasta.h"
#include <algorithm>
#include <list>
#include <set>

using namespace std;
using namespace vcflib;


void printSummary(char** argv) {
    cerr << "usage: " << argv[0] << " [options] [<vcf file>]" << endl
         << endl
         << "options:" << endl
         << "    -t, --truth-vcf FILE      use this VCF as ground truth for ROC generation" << endl
         << "    -w, --window-size N       compare records up to this many bp away (default 30)" << endl
         << "    -c, --complex             directly compare complex alleles, don't parse into primitives" << endl
         << "    -r, --reference FILE      FASTA reference file" << endl
         << endl
         << "Generates a pseudo-ROC curve using sensitivity and specificity estimated against" << endl
         << "a putative truth set.  Thresholding is provided by successive QUAL cutoffs." << endl;
    cerr << endl << "Type: statistics" << endl << endl;
    exit(0);
}

void buildVariantIntervalTree(VariantCallFile& variantFile,
                              map<string, IntervalTree<size_t, VariantLegacy*> >& variantIntervals,
                              list<VariantLegacy>& variants) {

    map<string, vector<Interval<size_t, VariantLegacy*> > > rawVariantIntervals;
    VariantLegacy var(variantFile);
    while (variantFile.getNextVariant(var)) {
        long int left = var.position;
        long int right = left + var.ref.size(); // this should be 1-past the end
        variants.push_back(var);
        VariantLegacy* v = &variants.back();
        rawVariantIntervals[var.sequenceName].push_back(Interval<size_t, VariantLegacy*>(left, right, v));
    }

    for (map<string, vector<Interval<size_t, VariantLegacy*> > >::iterator j = rawVariantIntervals.begin(); j != rawVariantIntervals.end(); ++j) {
        variantIntervals[j->first] = IntervalTree<size_t, VariantLegacy*>((vector<Interval<size_t, VariantLegacy*> >&&)j->second);
    }
}


void intersectVariant(VariantLegacy& var,
                      map<string, IntervalTree<size_t, VariantLegacy*> >& variantIntervals,
                      vector<string*>& commonAlleles,
                      vector<string*>& uniqueAlleles,
                      FastaReference& reference,
                      int windowsize = 50) {

    vector<Interval<size_t, VariantLegacy*> > results
        = variantIntervals[var.sequenceName].findContained(var.position - windowsize, var.position + var.ref.size() + windowsize);

    vector<VariantLegacy*> overlapping;

    for (vector<Interval<size_t, VariantLegacy*> >::iterator r = results.begin(); r != results.end(); ++r) {
        overlapping.push_back(r->value);
    }


    if (overlapping.empty()) {
        for (vector<string>::iterator a = var.alt.begin(); a != var.alt.end(); ++a) {
            uniqueAlleles.push_back(&*a);
        }
    } else {

        // get the min and max of the overlaps

        int haplotypeStart = var.position;
        int haplotypeEnd = var.position + var.ref.size();

        for (vector<VariantLegacy*>::iterator v = overlapping.begin(); v != overlapping.end(); ++v) {
            haplotypeStart = min((*v)->position, (long int) haplotypeStart);
            haplotypeEnd = max((*v)->position + (*v)->ref.size(), (long unsigned int) haplotypeEnd);
        }

        // for everything overlapping and the current variant, construct the local haplotype within the bounds
        // if there is an exact match, the allele in the current VCF does intersect

        string referenceHaplotype = reference.getSubSequence(var.sequenceName, haplotypeStart - 1, haplotypeEnd - haplotypeStart);
        map<string, vector<pair<VariantLegacy*, int> > > haplotypes; // map to variant and alt index

        for (vector<VariantLegacy*>::iterator v = overlapping.begin(); v != overlapping.end(); ++v) {
            VariantLegacy& variant = **v;
            int altindex = 0;
            for (vector<string>::iterator a = variant.alt.begin(); a != variant.alt.end(); ++a, ++altindex) {
                string haplotype = referenceHaplotype;
                // get the relative start and end coordinates for the variant alternate allele
                int relativeStart = variant.position - haplotypeStart;
                haplotype.replace(relativeStart, variant.ref.size(), *a);
                haplotypes[haplotype].push_back(make_pair(*v, altindex));
            }
        }


        // determine the non-intersecting alts
        for (vector<string>::iterator a = var.alt.begin(); a != var.alt.end(); ++a) {
            string haplotype = referenceHaplotype;
            int relativeStart = var.position - haplotypeStart;
            haplotype.replace(relativeStart, var.ref.size(), *a);
            map<string, vector<pair<VariantLegacy*, int> > >::iterator h = haplotypes.find(haplotype);
            if (h == haplotypes.end()) {
                uniqueAlleles.push_back(&*a);
            } else {
                commonAlleles.push_back(&*a);
            }
        }

    }
}


int main(int argc, char** argv) {

    string truthVcfFileName;
    string fastaFileName;
    bool complex = false;
    int windowsize = 30;

    if (argc == 1)
        printSummary(argv);

    int c;
    while (true) {
        static struct option long_options[] =
            {
                /* These options set a flag. */
                //{"verbose", no_argument,       &verbose_flag, 1},
                {"help", no_argument, 0, 'h'},
                {"window-size", required_argument, 0, 'w'},
                {"reference", required_argument, 0, 'r'},
                {"complex", required_argument, 0, 'c'},
                {"truth-vcf", required_argument, 0, 't'},
                {0, 0, 0, 0}
            };
        /* getopt_long stores the option index here. */
        int option_index = 0;

        c = getopt_long (argc, argv, "hcw:r:t:",
                         long_options, &option_index);

        if (c == -1)
            break;

        switch (c) {

	    case 'w':
            windowsize = atoi(optarg);
            break;

	    case 'r':
            fastaFileName = string(optarg);
            break;

	    case 't':
	        truthVcfFileName = optarg;
            break;

        case 'c':
            complex = true;
            break;

        case 'h':
            printSummary(argv);
            break;

        case '?':
            printSummary(argv);
            exit(1);
            break;

        default:
            abort ();
        }
    }

    VariantCallFile variantFile;
    bool usingstdin = false;
    string inputFilename;
    if (optind == argc - 1) {
        inputFilename = argv[optind];
        variantFile.open(inputFilename);
    } else {
        variantFile.open(std::cin);
        usingstdin = true;
    }

    if (!variantFile.is_open()) {
        cerr << "could not open VCF file" << endl;
        exit(1);
    }

    VariantCallFile truthVariantFile;
    if (!truthVcfFileName.empty()) {
        if (truthVcfFileName == "-") {
            if (usingstdin) {
                cerr << "cannot open both VCF file streams from stdin" << endl;
                exit(1);
            } else {
                truthVariantFile.open(std::cin);
            }
        } else {
            truthVariantFile.open(truthVcfFileName);
        }
        if (!truthVariantFile.is_open()) {
            cerr << "could not open VCF file " << truthVcfFileName << endl;
            exit(1);
        }
    }

    FastaReference reference;
    if (fastaFileName.empty()) {
        cerr << "a reference is required for the haplotype-based intersection used by vcfroc" << endl;
        exit(1);
    }
    reference.open(fastaFileName);

    // read the VCF file for union or intersection into an interval tree
    // indexed using some proximity window

    map<string, IntervalTree<size_t, VariantLegacy*> > truthVariantIntervals;
    list<VariantLegacy> truthVariants;
    buildVariantIntervalTree(truthVariantFile, truthVariantIntervals, truthVariants);

    map<string, IntervalTree<size_t, VariantLegacy*> > testVariantIntervals;
    list<VariantLegacy> testVariants;
    buildVariantIntervalTree(variantFile, testVariantIntervals, testVariants);

    map<long double, vector<VariantAllele*> > falseNegativeAllelesAtCutoff;  // false negative after this cutoff
    map<long double, vector<VariantAllele*> > falsePositiveAllelesAtCutoff;  // false positive until this cutoff
    list<VariantAllele*> allFalsePositiveAlleles;
    map<long double, vector<VariantAllele*> > allelesAtCutoff;
    //map<long double, vector<VariantAllele*> > totalAllelesAtCutoff;
    map<Variant*, map<string, vector<VariantAllele> > > parsedAlleles;
    map<long double, vector<VariantLegacy*> > callsByCutoff;

    // replicate this method, where Q is for each unique Q in the set
    //vcfintersect -r $reference -v -i $results.$Q.vcf $answers_primitives | vcfstats >false_negatives.$Q.stats
    //vcfintersect -r $reference -v -i $answers_primitives $results.$Q.vcf | vcfstats >false_positives.$Q.stats

    for (list<VariantLegacy>::iterator v = testVariants.begin(); v != testVariants.end(); ++v) {
        // TODO allow different cutoff sources
        callsByCutoff[v->quality].push_back(&*v);
    }

    // add false negatives at any cutoff
    for (list<VariantLegacy>::iterator v = truthVariants.begin(); v != truthVariants.end(); ++v) {
        VariantLegacy& variant = *v;
        vector<string*> commonAlleles;
        vector<string*> uniqueAlleles;
        intersectVariant(variant, testVariantIntervals,
                         commonAlleles, uniqueAlleles, reference);
        if (complex) {
            parsedAlleles[&*v] = variant.flatAlternates();
        } else {
            parsedAlleles[&*v] = variant.legacy_parsedAlternates();
        }
        // unique alleles are false negatives regardless of cutoff
        for (vector<string*>::iterator a = uniqueAlleles.begin(); a != uniqueAlleles.end(); ++a) {
            vector<VariantAllele>& alleles = parsedAlleles[&*v][**a];
            for (vector<VariantAllele>::iterator va = alleles.begin(); va != alleles.end(); ++va) {
                if (va->ref != va->alt) { 		// use only non-reference alleles
                    // false negatives at threshold 0 XXX --- may not apply if threshold is generalized
                    falseNegativeAllelesAtCutoff[-1].push_back(&*va);
                }
            }
        }
    }

    for (map<long double, vector<VariantLegacy*> >::iterator q = callsByCutoff.begin(); q != callsByCutoff.end(); ++q) {
        long double threshold = q->first;
        vector<VariantLegacy*>& variants = q->second;
        for (vector<VariantLegacy*>::iterator v = variants.begin(); v != variants.end(); ++v) {
            VariantLegacy& variant = **v;
            vector<string*> commonAlleles;
            vector<string*> uniqueAlleles;
            intersectVariant(variant, truthVariantIntervals,
                             commonAlleles, uniqueAlleles, reference);
            if (complex) {
                parsedAlleles[*v] = variant.flatAlternates();
            } else {
                parsedAlleles[*v] = variant.legacy_parsedAlternates();
            }

            map<string, vector<VariantAllele> >& parsedAlts = parsedAlleles[*v];
            // push VariantAllele*'s into the FN and FP alleles at cutoff vectors
            for (vector<string*>::iterator a = commonAlleles.begin(); a != commonAlleles.end(); ++a) {
                vector<VariantAllele>& alleles = parsedAlleles[*v][**a];
                for (vector<VariantAllele>::iterator va = alleles.begin(); va != alleles.end(); ++va) {
                    if (va->ref != va->alt) { 		// use only non-reference alleles
                        allelesAtCutoff[threshold].push_back(&*va);
                        falseNegativeAllelesAtCutoff[threshold].push_back(&*va);
                    }
                }
            }
            for (vector<string*>::iterator a = uniqueAlleles.begin(); a != uniqueAlleles.end(); ++a) {
                vector<VariantAllele>& alleles = parsedAlts[**a];
                for (vector<VariantAllele>::iterator va = alleles.begin(); va != alleles.end(); ++va) {
                    if (va->ref != va->alt) { 		// use only non-reference alleles
                        allelesAtCutoff[threshold].push_back(&*va);
                        allFalsePositiveAlleles.push_back(&*va);
                        falsePositiveAllelesAtCutoff[threshold].push_back(&*va);
                    }
                }
            }
        }
    }


    // output results
    int totalSNPs = 0;
    int falsePositiveSNPs = 0;
    int falseNegativeSNPs = 0;
    int totalIndels = 0;
    int falsePositiveIndels = 0;
    int falseNegativeIndels = 0;
    int totalComplex = 0;
    int falsePositiveComplex = 0;
    int falseNegativeComplex = 0;

    // write header

    cout << "threshold" << "\t"
         << "num_snps" << "\t"
         << "false_positive_snps" << "\t"
         << "false_negative_snps" << "\t"
         << "num_indels" << "\t"
         << "false_positive_indels" << "\t"
         << "false_negative_indels" << "\t"
         << "num_complex" << "\t"
         << "false_positive_complex" << "\t"
         << "false_negative_complex" << endl;

    // count total alleles in set
    for (map<long double, vector<VariantAllele*> >::iterator a = allelesAtCutoff.begin(); a != allelesAtCutoff.end(); ++a) {
        vector<VariantAllele*>& alleles = a->second;
        for (vector<VariantAllele*>::iterator va = alleles.begin(); va != alleles.end(); ++va) {
            VariantAllele& allele = **va;
            if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
                ++totalSNPs;
            } else if (allele.ref.size() != allele.alt.size()) {
                if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                    ++totalIndels;
                } else {
                    ++totalComplex;
                }
            } else {
                ++totalComplex;
            }
        }
    }

    // tally total false positives
    for (list<VariantAllele*>::iterator va = allFalsePositiveAlleles.begin(); va != allFalsePositiveAlleles.end(); ++va) {
        VariantAllele& allele = **va;
        if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
            ++falsePositiveSNPs;
        } else if (allele.ref.size() != allele.alt.size()) {
            if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                ++falsePositiveIndels;
            } else {
                ++falsePositiveComplex;
            }
        } else {
            ++falsePositiveComplex;
        }
    }

    // get categorical false negatives
    vector<VariantAllele*>& categoricalFalseNegatives = falseNegativeAllelesAtCutoff[-1];
    for (vector<VariantAllele*>::iterator va = categoricalFalseNegatives.begin(); va != categoricalFalseNegatives.end(); ++va) {
        VariantAllele& allele = **va;
        if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
            assert(allele.ref.size() == 1);
            ++falseNegativeSNPs;
        } else if (allele.ref.size() != allele.alt.size()) {
            if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                ++falseNegativeIndels;
            } else {
                ++falseNegativeComplex;
            }
        } else {
            ++falseNegativeComplex;
        }
    }
    cout << -1 << "\t"
         << totalSNPs << "\t"
         << falsePositiveSNPs << "\t"
         << falseNegativeSNPs << "\t"
         << totalIndels << "\t"
         << falsePositiveIndels << "\t"
         << falseNegativeIndels << "\t"
         << totalComplex << "\t"
         << falsePositiveComplex << "\t"
         << falseNegativeComplex << endl;

    for (map<long double, vector<VariantAllele*> >::iterator a = allelesAtCutoff.begin(); a != allelesAtCutoff.end(); ++a) {
        vector<VariantAllele*>& alleles = a->second;
        long double threshold = a->first;
        for (vector<VariantAllele*>::iterator va = alleles.begin(); va != alleles.end(); ++va) {
            VariantAllele& allele = **va;
            if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
                assert(allele.ref.size() == 1);
                --totalSNPs;
            } else if (allele.ref.size() != allele.alt.size()) {
                if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                    --totalIndels;
                } else {
                    --totalComplex;
                }
            } else {
                --totalComplex;
            }
        }
        vector<VariantAllele*>& falseNegatives = falseNegativeAllelesAtCutoff[threshold];
        for (vector<VariantAllele*>::iterator va = falseNegatives.begin(); va != falseNegatives.end(); ++va) {
            VariantAllele& allele = **va;
            if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
                assert(allele.ref.size() == 1);
                ++falseNegativeSNPs;
            } else if (allele.ref.size() != allele.alt.size()) {
                if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                    ++falseNegativeIndels;
                } else {
                    ++falseNegativeComplex;
                }
            } else {
                ++falseNegativeComplex;
            }
        }
        vector<VariantAllele*>& falsePositives = falsePositiveAllelesAtCutoff[threshold];
        for (vector<VariantAllele*>::iterator va = falsePositives.begin(); va != falsePositives.end(); ++va) {
            VariantAllele& allele = **va;
            if (allele.ref.size() == 1 && allele.ref.size() == allele.alt.size()) {
                assert(allele.ref.size() == 1);
                --falsePositiveSNPs;
            } else if (allele.ref.size() != allele.alt.size()) {
                if (allele.ref.size() == 1 || allele.alt.size() == 1) {
                    --falsePositiveIndels;
                } else {
                    --falsePositiveComplex;
                }
            } else {
                --falsePositiveComplex;
            }
        }
        cout << threshold << "\t"
             << totalSNPs << "\t"
             << falsePositiveSNPs << "\t"
             << falseNegativeSNPs << "\t"
             << totalIndels << "\t"
             << falsePositiveIndels << "\t"
             << falseNegativeIndels << "\t"
             << totalComplex << "\t"
             << falsePositiveComplex << "\t"
             << falseNegativeComplex << endl;

    }

    exit(0);  // why?
    return 0;

}
